Infinite recursion error when using $ to consume entire input type

A change in 2.2.0 broke type inference for non-primitive types, even when using the new `$$` syntax to force a required input. I did some basic digging into the generated file, and it seems to be linked to the `VariabledInput` changes.

**Proposed Fix**: Add `Variable<INPUT_TYPE,any>` to the generated input types, allowing an entire input type to be consumed as a Variable.

Reproduction:

1. Hasura style graphql schema (any works)
2. Construct a mutation that takes an entire input type as a variable

```ts
const upsertJobMutation = mutation((m) => [
  m.insert_job_one(
    {
      object: $$("job")
    },
    (j) => [j.id]
  ),
]);
```

Generated code:
```ts
/**
 * insert a single row into the table: "jobs"
 */
insert_job_one<Args extends VariabledInput<{
  object: job_insert_input
  on_conflict?: job_on_conflict | null | undefined,
}>,Sel extends Selection<job>>(args: Args, selectorFn: (s: job) => [...Sel]):$Field<"insert_job_one", GetOutput<Sel> | undefined , GetVariables<Sel, Args>> {

  const options = {
    argTypes: {
          object: "job_insert_input!",
          on_conflict: "job_on_conflict"
        },
    args,

    selection: selectorFn(new job)
  };
  return this.$_select("insert_job_one", options) as any
}
```

Working Code:
```ts
/**
 * insert a single row into the table: "jobs"
 */
insert_job_one<Args extends VariabledInput<{
  object: job_insert_input | Variable<job_insert_input, any>
  on_conflict?: job_on_conflict | null | undefined  | Variable<job_on_conflict, any>,
}>,Sel extends Selection<job>>(args: Args, selectorFn: (s: job) => [...Sel]):$Field<"insert_job_one", GetOutput<Sel> | undefined , GetVariables<Sel, Args>> {

  const options = {
    argTypes: {
          object: "job_insert_input!",
          on_conflict: "job_on_conflict"
        },
    args,

    selection: selectorFn(new job)
  };
  return this.$_select("insert_job_one", options) as any
}
```

Expected: Can consume `$$("job")` as a matching input of `job_insert_input`

Actual: TS error `Type instantiation is excessively deep and possibly infinite.ts(2589)`

I found the reason for the issue. A trick I use to debug these is to either remove the recursive input type to see the issue, or to switch to a limited-depth definition (but that one doesn't support variables past that depth)

```typescript
type Decrement<X extends number> = X extends 5
  ? 4
  : X extends 4
  ? 3
  : X extends 3
  ? 2
  : X extends 2
  ? 1
  : 0

type VariabledInput<T, MaxDepth extends number = 5> = MaxDepth extends 0
  ? T
  : [T] extends [$Atomic | null | undefined]
  ? Variable<T, any> | T
  : T extends ReadonlyArray<infer I>
  ? Variable<T, any> | T | ArrayInput<I>
  : T extends Record<string, any>
  ? { [K in keyof T]: VariabledInput<T[K], Decrement<MaxDepth>> } | T
  : never
```

It turned out that when `T extends Record<string, any>`, we don't allow Variable<T, any> as the option anymore. Should be a one-line fix.

I wonder if it makes sense to limit the recursion depth to get better error messages? When I limit it to 5, I get the following error:

```
examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | { ...; })[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: ...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_and' circularly references itself in mapped type '{ _and?: any; _not?: any; _or?: any; name?: unknown; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.
examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_not' circularly references itself in mapped type '{ _and?: any; _not?: any; _or?: any; name?: unknown; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_or' circularly references itself in mapped type '{ [K in keyof booking_channel_bool_exp]: ExtractInputVariables<booking_channel_bool_exp[K]>; }'.

examples/test-x.good.ts(46,3): error TS2615: Type of property '_or' circularly references itself in mapped type '{ [K in keyof { _and?: readonly booking_channel_bool_exp[] | Variable<readonly booking_channel_bool_exp[], any> | readonly (booking_channel_bool_exp | ...)[] | undefined; _not?: booking_channel_bool_exp | ... 1 more ... | undefined; _or?: readonly booking_channel_bool_exp[] | ... 2 more ... | undefined; name?: Strin...'.

examples/test-x.good.ts(48,7): error TS2322: Type 'Variable<{}, "job">' is not assignable to type 'booking_channel_insert_input | { name?: string | Variable<string | null | undefined, any> | null | undefined; }'.
```

Notice the last message - when the recursion depth is limited, we also get the actual error at the end.

--- a/package.json
+++ b/package.json
@@ -32,7 +32,8 @@
     "test:watch": "tap -w test/**/*.test.ts -R spec",
     "regenerate:big-schema": "swc-node src/cli.ts --schema examples/x.graphql --output examples/x.ts",
     "regenerate:small-schema": "swc-node src/cli.ts --schema examples/zeus.graphql --output examples/zeus.ts",
-    "regenerate:country-schema": "swc-node src/cli.ts --schema examples/countries-schema.graphql --output examples/countries-schema.ts"
+    "regenerate:country-schema": "swc-node src/cli.ts --schema examples/countries-schema.graphql --output examples/countries-schema.ts",
+    "prepublish": "yarn build && yarn test"
   },
   "targets": {
     "main": {

--- a/package.json
+++ b/package.json
@@ -33,7 +33,7 @@
     "regenerate:big-schema": "swc-node src/cli.ts --schema examples/x.graphql --output examples/x.ts",
     "regenerate:small-schema": "swc-node src/cli.ts --schema examples/zeus.graphql --output examples/zeus.ts",
     "regenerate:country-schema": "swc-node src/cli.ts --schema examples/countries-schema.graphql --output examples/countries-schema.ts",
-    "prepublish": "yarn build && yarn test"
+    "prepublishOnly": "yarn build && yarn test"
   },
   "targets": {
     "main": {

--- a/src/preamble.src.ts
+++ b/src/preamble.src.ts
@@ -31,7 +31,7 @@ type VariabledInput<T> = [T] extends [$Atomic | null | undefined]
   : T extends ReadonlyArray<infer I>
   ? Variable<T, any> | T | ArrayInput<I>
   : T extends Record<string, any>
-  ? { [K in keyof T]: VariabledInput<T[K]> } | T
+  ? Variable<T, any> | { [K in keyof T]: VariabledInput<T[K]> } | T
   : never
 
 type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void

--- a/test/examples/test-x.good.ts
+++ b/test/examples/test-x.good.ts
@@ -1,7 +1,7 @@
 // Todo: add big query tests
 
 import { verify } from './verify'
-import { query, order_by, $, $$ } from './x.graphql.api'
+import { query, order_by, $, $$, mutation } from './x.graphql.api'
 
 let orderByTest = query(q => [
   q.bookings(
@@ -39,6 +39,15 @@ let bookingsBetween = query(q => [
         c.createdAt,
       ]),
     ]
+  ),
+])
+
+const upsertBookingChannelMutation = mutation(m => [
+  m.insert_booking_channel_one(
+    {
+      object: $$('bc'),
+    },
+    b => [b.name]
   ),
 ])
 
@@ -81,5 +90,10 @@ export default [
     },
     schemaPath: 'x.graphql',
     string: bookingsBetweenString,
+  }),
+  verify({
+    query: upsertBookingChannelMutation,
+    variables: { bc: { name: 'hello' } },
+    schemaPath: 'x.graphql',
   }),
 ]

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -89,6 +89,7 @@ for (let schema of glob.sync(`./examples/*.graphql`, { cwd: __dirname })) {
       t.test(`compiles with example ${exampleName}`, async t => {
         let res = compileTs(example)
         if (res.status) {
+          console.error(res.stdout.toString())
           t.fail(res.stdout.toString())
         }
       })

